tools/traffic/test/blob_writer_test.go

```diff
@@ -0,0 +1,143 @@
+package test
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"github.com/Layr-Labs/eigenda/common"
+	tu "github.com/Layr-Labs/eigenda/common/testutils"
+	"github.com/Layr-Labs/eigenda/encoding/utils/codec"
+	"github.com/Layr-Labs/eigenda/tools/traffic/config"
+	"github.com/Layr-Labs/eigenda/tools/traffic/metrics"
+	"github.com/Layr-Labs/eigenda/tools/traffic/workers"
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/exp/rand"
+	"sync"
+	"testing"
+	"time"
+)
+
+func TestBlobWriter(t *testing.T) {
+	tu.InitializeRandom()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	waitGroup := sync.WaitGroup{}
+	logger, err := common.NewLogger(common.DefaultLoggerConfig())
+	assert.Nil(t, err)
+	startTime := time.Unix(rand.Int63()%2_000_000_000, 0)
+	ticker := newMockTicker(startTime)
+
+	dataSize := rand.Uint64()%1024 + 64
+
+	authenticated := rand.Intn(2) == 0
+	var signerPrivateKey string
+	if authenticated {
+		signerPrivateKey = "asdf"
+	}
+
+	randomizeBlobs := rand.Intn(2) == 0
+
+	useCustomQuorum := rand.Intn(2) == 0
+	var customQuorum []uint8
+	if useCustomQuorum {
+		customQuorum = []uint8{1, 2, 3}
+	}
+
+	config := &config.WorkerConfig{
+		DataSize:         dataSize,
+		SignerPrivateKey: signerPrivateKey,
+		RandomizeBlobs:   randomizeBlobs,
+		CustomQuorums:    customQuorum,
+	}
+
+	lock := sync.Mutex{}
+
+	disperserClient := newMockDisperserClient(t, &lock, authenticated)
+	unconfirmedKeyHandler := newMockKeyHandler(t, &lock)
+
+	generatorMetrics := metrics.NewMockMetrics()
+
+	writer := workers.NewBlobWriter(
+		&ctx,
+		&waitGroup,
+		logger,
+		ticker,
+		config,
+		disperserClient,
+		unconfirmedKeyHandler,
+		generatorMetrics)
+	writer.Start()
+
+	errorProbability := 0.1
+	errorCount := 0
+
+	var previousData []byte
+
+	for i := 0; i < 100; i++ {
+		if rand.Float64() < errorProbability {
+			disperserClient.DispenseErrorToReturn = fmt.Errorf("intentional error for testing purposes")
+			errorCount++
+		} else {
+			disperserClient.DispenseErrorToReturn = nil
+		}
+
+		// This is the key that will be assigned to the next blob.
+		disperserClient.KeyToReturn = make([]byte, 32)
+		_, err = rand.Read(disperserClient.KeyToReturn)
+		assert.Nil(t, err)
+
+		// Move time forward, allowing the writer to attempt to send a blob.
+		ticker.Tick(1 * time.Second)
+
+		// Wait until the writer finishes its work.
+		tu.AssertEventuallyTrue(t, func() bool {
+			lock.Lock()
+			defer lock.Unlock()
+			return int(disperserClient.DisperseCount) > i && int(unconfirmedKeyHandler.Count)+errorCount > i
+		}, time.Second)
+
+		// These methods should be called exactly once per tick if there are no errors.
+		// In the presence of errors, nothing should be passed to the unconfirmed key handler.
+		assert.Equal(t, uint(i+1), disperserClient.DisperseCount)
```

Comment:
Timing/synchronizing with another goroutine might be tricky just using time. Does this reliably pass? How do you know if L101 will execute after disperser request has been made? 
Maybe one thing we could do is make `run` method public and run test against this method? That way we can assert expectations after the method returns.

Reply:
Empirically, this was reliably passing (ran it in a loop a few thousand times). Primary reason why I was going through all the trouble of doing it in an attempt to put the tests in another directory without making methods public that shouldn't be public.

I understand your general concern though. I refactored this so that I don't rely on timing. As a result, the test files will end up being in the same package as the other files. Ugly to me, but it seems like a fairly normal practice in go.